plugin/keyring_vault/vault_curl.cc

```diff
@@ -0,0 +1,177 @@
+#include <algorithm>
+#include "vault_curl.h"
+#include "base64.h"
+
+namespace keyring
+{
+
+static size_t write_response_memory(void *contents, size_t size, size_t nmemb, void *userp)
+{
+  size_t realsize = size * nmemb;
+  std::stringstream *read_data = (std::stringstream *)userp;
+
+  read_data->write((char*)contents, realsize);
+  if (!read_data->good())
+    return 0;
+  return realsize;
+}
+
+std::string Vault_curl::get_error_from_curl(CURLcode curl_code)
+{
+  size_t len = strlen(curl_errbuf);
+  std::stringstream ss;
+  if (curl_code != CURLE_OK)
+  {
+    ss << "Curl returned this error code: " << curl_code;
+    ss << " with error message : ";
+    if(len)
+      ss << curl_errbuf;
+    else
+      ss << curl_easy_strerror(curl_code);
+  }
+  return ss.str();
+}
+
+my_bool Vault_curl::init(Vault_credentials *vault_credentials)
+{
+  curl = curl_easy_init();
+  if (curl == NULL)
+  {
+    logger->log(MY_ERROR_LEVEL, "Could not create CURL session");
+    return TRUE;
+  }
+  this->token_header = "X-Vault-Token:" + (*vault_credentials)["token"];
+  this->vault_url = (*vault_credentials)["vault_url"] + "/v1/" + (*vault_credentials)["secret_mount_point"];
+  this->vault_ca = (*vault_credentials)["vault_ca"];
+  if(this->vault_ca.empty())
+  {
+    logger->log(MY_WARNING_LEVEL, "There is no vault_ca specified in keyring_vault's configuration file. "
+                                  "Please make sure that Vault's CA certificate is trusted by the machine from "
+                                  "which you intend to connect to Vault."); 
+  }
+  return FALSE;
+}
+
+my_bool Vault_curl::reset_curl_session()
+{
+  CURLcode curl_res = CURLE_OK;
+  curl_easy_reset(curl);
+  read_data_ss.str("");
+  read_data_ss.clear();
+  curl_errbuf[0] = '\0';
+  if (list != NULL)
+  {
+    curl_slist_free_all(list);
+    list = NULL;
+  }
+
+  if ((list = curl_slist_append(list, token_header.c_str())) == NULL ||
+      (list = curl_slist_append(list, "Content-Type: application/json")) == NULL ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_errbuf)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_response_memory)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_WRITEDATA, reinterpret_cast<void*>(&read_data_ss))) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 1)) != CURLE_OK ||
+      (curl_res = curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L)) != CURLE_OK ||
+      (vault_ca.empty() == false &&
```

Comment:
here and everywhere: do not compare boolean expressions with true/false
it's either `expr` or `!expr`